### PR TITLE
keystore abstraction and adaption betweeen Loopp and Caigo

### PR DESCRIPTION
**Pull request by krehermann**
Adds Keystore abstraction to the transaction manager

Most of the work is translating from the LOOPp interface definition which is raw bytes to the `caigo` keystore definition which is big.Ints.

In order to achieve that translation, we introduce a Keystore Adapter  that implements both Keystore interfaces and leverages an intermediate representation of the signature to handle the roundtrip between the formats.

### CODE REVIEW

File: relayer/pkg/chainlink/keys/keystore.go
Comment (aalu1418) [inline relayer/pkg/chainlink/keys/keystore.go]: noted in other PR - but having a context here would be useful so it doesn't block forever if something breaks :)

File: relayer/pkg/chainlink/keys/keystore.go
Comment (aalu1418) [inline relayer/pkg/chainlink/keys/keystore.go]: just curious: is this for testing or something that will be a part of the codebase?

**Review by aalu1418 — COMMENTED** [review]
✅ 

overall direction looks good to me!

File: relayer/pkg/chainlink/keys/keystore.go
Comment (jmank88) [inline relayer/pkg/chainlink/keys/keystore.go]: Can these ints be negative? b/c `Bytes()` doesn't preserve the sign. I ran in to this with LOOPs and ended up using a protobuf https://github.com/smartcontractkit/chainlink-relay/blob/main/pkg/loop/internal/pb/big.go

File: relayer/pkg/chainlink/keys/keystore.go
Comment (krehermann) [reply relayer/pkg/chainlink/keys/keystore.go]: yes, this was test code. reorganized now

File: relayer/pkg/chainlink/keys/keystore.go
Comment (krehermann) [reply relayer/pkg/chainlink/keys/keystore.go]: done, thanks for catching

File: relayer/pkg/chainlink/keys/keystore.go
Comment (krehermann) [reply relayer/pkg/chainlink/keys/keystore.go]: that's a good question. 

yes, based on the elliptic curve, they can be negative.

the current OCR key signing for starknet is using big.Int.Bytes, and i largely copied that.

@aalu1418 what is the expectation regarding the sign of the points on the curve -- is the OCR implementation consciously choosing to ignore/drop it, or is that an oversight in the implementation? 

https://github.com/smartcontractkit/chainlink-starknet/blob/develop/relayer/pkg/chainlink/keys/ocr2key.go

File: relayer/go.sum
Comment (aalu1418) [inline relayer/go.sum]: nit: probs need a tidy here 😄

File: relayer/pkg/chainlink/keys/keystore.go
Comment (aalu1418) [reply relayer/pkg/chainlink/keys/keystore.go]: for the OCR key - i remember that being a conscious decision to only use the positive values

however, for the transmitter key/signatures I'm not sure - since we have previously passed the private key caigo handled the signs / encodings internally 🤔

Comment (krehermann) [reply relayer/pkg/chainlink/keys/keystore.go]: i looked at the official python implementation. in my reading, all the int values are positive. can one of you read the python and/or caigo give your assessment? cc @archseer 

i'm also going to ask research on the general topic of big.Int.Set/Bytes and elliptic curves.

the private key is positve     https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L103

the msgHash is positive https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L139


the signature components --
r is positive:         https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L156

s is positive: 

https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L170

https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L111

https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/math_utils.py#L52

Comment (jmank88) [reply relayer/pkg/chainlink/keys/keystore.go]: Would it make sense to sanity check and panic if negative?

Comment (krehermann) [reply relayer/pkg/chainlink/keys/keystore.go]: yes, good idea. i think the right place for that is the `Decode` implementation, which i've now added

File: integration-tests/go.mod
Comment (jmank88) [inline integration-tests/go.mod]:
```suggestion
```

File: relayer/pkg/chainlink/keys/nonce_test.go
Comment (jmank88) [inline relayer/pkg/chainlink/keys/nonce_test.go]:
```suggestion
```

File: ops/go.mod
Comment (jmank88) [inline ops/go.mod]: unresolved conflicts

File: relayer/pkg/chainlink/keys/keystore.go
Comment (jmank88) [reply relayer/pkg/chainlink/keys/keystore.go]: Where is the sanity check? I see `(s *signature) bytes()` checking length, but not sign.

**Comment by jmank88** [comment]
We'd like to wait to merge until after #284 (reverts breaking changes), so we can get this in to core ahead of those breaking changes.
Edit: Merged! Sorry for the conflicts :facepalm:

File: relayer/pkg/chainlink/keys/keystore_test.go
Comment (jmank88) [inline relayer/pkg/chainlink/keys/keystore_test.go]: WDYT about trying to avoid the label somehow?
```suggestion
	generate := func() bool {
		select {
		case <-timer.C:
			key = nil
			err = fmt.Errorf("failed to generate test key in allotted time")
			return true
		default:
			key, err = caigo.Curve.GetRandomPrivateKey()
			if err == nil {
				return true
			}
		}
		return false
	}
	for !generate() { }
```

Comment (krehermann) [reply relayer/pkg/chainlink/keys/keystore_test.go]: sure.

what is the objection about the named loop?

File: integration-tests/go.mod
Comment (jmank88) [reply integration-tests/go.mod]: There's still an odd separate `// indirect` block here.

File: relayer/pkg/chainlink/keys/keystore_test.go
Comment (jmank88) [reply relayer/pkg/chainlink/keys/keystore_test.go]: Just that they are generally more difficult to reason about and prone to errors in future edits. Or sometimes a code smell that there is too much nesting/indentation regardless.
I usually find alternative styles to be simpler anyways, except maybe in cases of tight double for loops?

File: relayer/pkg/chainlink/keys/keystore.go
Comment (jmank88) [inline relayer/pkg/chainlink/keys/keystore.go:118]: Does starknet have a balance monitor? That's what this was supporting.

Comment (krehermann) [reply relayer/pkg/chainlink/keys/keystore.go:118]: @aalu1418?

the keystore interface that had been used in this repo doesn't need this func as far as i can tell and i didn't find anything like a balance monitor

Comment (aalu1418) [reply relayer/pkg/chainlink/keys/keystore.go:118]: we don't have one yet, but that's usually one of the first features NOPs request

File: relayer/pkg/chainlink/txm/txm_test.go
Comment (jmank88) [inline relayer/pkg/chainlink/txm/txm_test.go]: Drop or restore?

File: integration-tests/go.mod
Comment (krehermann) [reply integration-tests/go.mod]: i thought it would have been fixed magically my one of my many go mod tidy incantations. surprised that it wasn't. should be fixed now

File: relayer/pkg/chainlink/txm/txm_test.go
Comment (krehermann) [reply relayer/pkg/chainlink/txm/txm_test.go]: are you getting pinged to review this over and over? i've been battling the code all day and it's been WIP. i thought you would only get pinged if i manually request review.

going forward would you prefer i put it back in draft when i'm pushing a lot of changes that i don't expect to be final?

File: integration-tests/go.mod
Comment (jmank88) [reply integration-tests/go.mod]: yeah the tooling doesn't undo these, and it occasionally introduces them for some reason :confused:

File: relayer/pkg/chainlink/txm/txm_test.go
Comment (jmank88) [reply relayer/pkg/chainlink/txm/txm_test.go]: NBD I was just trying to stay ahead since we are blocked on the error signature change due to an odd testing module dependency between this repo and solana.
Draft could still be helpful though :shrug:

File: relayer/pkg/chainlink/keys/keystore.go
Comment (jmank88) [reply relayer/pkg/chainlink/keys/keystore.go]: bump ^

File: relayer/pkg/chainlink/txm/txm.go
Comment (jmank88) [inline relayer/pkg/chainlink/txm/txm.go]: What does this comment mean?

Comment (krehermann) [reply relayer/pkg/chainlink/txm/txm.go]: it's WIP garbage that needs to be deleted. thanks for finding it

File: relayer/pkg/chainlink/keys/keystore.go
Comment (jmank88) [inline relayer/pkg/chainlink/keys/keystore.go]: Looking forward towards build separation (or even just our current blocking test entanglement), I think this type will be problematic because it can't be shared from both sides. We would need to have it represented in `chainlink-relay` to do that. Is it important for this to be a custom encoding? Or could we reuse the existing protobuf that already supports the reporting interfaces? https://github.com/smartcontractkit/chainlink-relay/blob/main/pkg/loop/internal/pb/relayer.proto#L306

File: relayer/pkg/chainlink/keys/keystore.go
Comment (krehermann) [reply relayer/pkg/chainlink/keys/keystore.go]: Oh right, i meant to reply to this.

It turns out that this is not needed. it's not possible to check that they are negative -- the SetBytes assumes an absolute value, so you would need a different en/decoding to represent a negative number and therefore you can't even check for negatives in a meaningful way here.

the same idea applies to the bytes() func: since it's using big.Int.Bytes(), it is not possible to encode negative values.

the only place a negative could be represented is in line 110-112. but there again the value is coming from the caigo.Sign, which returns positive values. 

we could add a panic there i suppose. wdyt?

i've written tests for the round trip in keystore and signature. do you see anything missing in them?

File: relayer/pkg/chainlink/keys/keystore.go
Comment (krehermann) [reply relayer/pkg/chainlink/keys/keystore.go]: this is private and isn't shared, unless i'm misunderstanding you.

it's only needed in this repo and more specifically, only for the txm.

it's helper layer that is used in a specific implementation of the keystore adaption logic. in principle the keystore adapter is an interface and another implementation could be used, but there are a lot of constraints that have to be met, and this is private way to enforce those constraints for a specific implementation.

the protobuf definition doesn't help here because we need an encoding that turns bytes into multiple big.Ints. we would change the problem into needing an encoding for multiple protobuf instances

File: relayer/pkg/chainlink/keys/keystore.go
Comment (jmank88) [reply relayer/pkg/chainlink/keys/keystore.go]: If `(s *signature) bytes()` is called with a signature that has negative values, it will silently drop that information. IMHO this is where the panic should occur: at the narrowest scope without making assumptions about anything else.

File: relayer/pkg/chainlink/keys/keystore.go
Comment (krehermann) [reply relayer/pkg/chainlink/keys/keystore.go]: more generally, we need a good story for this kind of pattern.

there is no constraint on new chains that they have to accept bytes as a signature and therefore there will be a need to adapt the loop.keystore interface to chain-specific logic

Comment (jmank88) [reply relayer/pkg/chainlink/keys/keystore.go]: > this is private and isn't shared, unless i'm misunderstanding you.

We need a way to understand the encoding on both ends. It would be better not to have separate redundant/duplicate code that needs to be kept in sync.

> we would change the problem into needing an encoding for multiple protobuf instances

Yes, but that is a trivial problem.

> more generally, we need a good story for this kind of pattern.
>
> there is no constraint on new chains that they have to accept bytes as a signature and therefore there will be a need to adapt the loop.keystore interface to chain-specific logic

This is why I suggested using a dynamic protobuf message instead of raw `[]byte`.

Comment (jmank88) [reply relayer/pkg/chainlink/keys/keystore.go]: I think https://protobuf.dev/programming-guides/proto3/#oneof is probably the idiomatic way to support different types of signatures here :shrug:

Comment (krehermann) [reply relayer/pkg/chainlink/keys/keystore.go]: >We need a way to understand the encoding on both ends. It would be better not to have separate redundant/duplicate code that needs to be kept in sync.

I don't think the explicit encoding needs to be exposed, rather we need a consistent use of the implemented interfaces.

the encoding is hidden behind the interfaces. if we use consistent implementations on both sides then only the implementors/owners of the chain need to think about this problem.

i dislike changing the protobuf defintion because it gets into a gray area of leaking the implementation details about a specific chains.

if a 3rd chain comes along with some other signature, we would need to update the proto rather to support it. that is a lot of overhead and coupling. instead we can leave it to the implementors to encode and decode the bytes as they need for there use case

it feels like the same kind of problem as marshaling and unmarshaling json or http -- as an implementor you need to know how to interpret the bytes, but the protocol that transports them doesn't care

Comment (jmank88) [reply relayer/pkg/chainlink/keys/keystore.go]: > if we use consistent implementations on both sides then only the implementors/owners of the chain need to think about this problem.

The implementation must exist in chainlink-relay, or be duplicated.

> i dislike changing the protobuf defintion because it gets into a gray area of leaking the implementation details about a specific chains.

The details are already leaked. This is just a way of standardizing them.

> if a 3rd chain comes along with some other signature, we would need to update the proto rather to support it. that is a lot of overhead and coupling. instead we can leave it to the implementors to encode and decode the bytes as they need for there use case

Core already needs to be updated, so a new relay version is not much more IMHO.

> it feels like the same kind of problem as marshaling and unmarshaling json or http -- as an implementor you need to know how to interpret the bytes, but the protocol that transports them doesn't care

The protocol is not the issue here. Core cares and it cannot import the chainlink-starknet.

Comment (krehermann) [reply relayer/pkg/chainlink/keys/keystore.go]: updated to use common code from the chainlink-relay repo. please review again

**Review by jmank88 — DISMISSED** [review]

**Review by jmank88 — DISMISSED** [review]

**Review by jmank88 — APPROVED** [review]

**Pull request merged**
